### DISCUSSION TITLE
How to improve T4 code generation performance

### DISCUSSION THREAD

**Discussion by dipique** [linq2db/linq2db#3681] [Q&A] [answered]
I implemented linq2db for my team a couple years back. The response has been really positive since we ripped out an ancient, poorly-maintained EF6 implementation. At first, T4 generation was really fast (I limited the schema list to keep it light), but every time we add a new schema, refreshing the generated file gets a little slower and by this time, refreshing (from a local Sql Server instance) completely freezes visual studio for several minutes. That makes it a lot harder to convince people to keep it up to date.

Perhaps there are general performance best practices that would help--and if so, I'm all ears. But I was thinking that a lot of times we're just changing one column or table or schema. And it would be amazing if we could somehow use that information to do more targeted updates. Is there a setup or tool that would allow us to do something like that? I'd like to tell linq2db, "I'm adding a `ContactId` field to `Order.Customer` which is a FK to `User.Contact`; make that change and any related changes (e.g. adding a navigation backreference to `User.Contact`)".

As I understand it, the underlying barrier is that the metadata is not stored once it is gathered, and so it has to be fully re-gathered every single time. Is that a fair assessment? If so, what if that metadata were stored (serialized?) and kept in source control (e.g. `MyDb.metadata.json`)--or maybe even the generated file itself is enough to rebuild the metadata?--and then a special T4 template could be used that specifies a specific table or schema to refresh, after which the actual model generation could happen normally. Is that something I could set up without tearing the architecture apart?

What gives me hope is that I've noticed that sometimes refreshing the model a second time runs much, much faster. If there's any way to replicate those conditions as needed it would be an enormous benefit.

Sorry this was so long; any guidance would be appreciated.

**Comment by MaceWindu** [comment, marked as answer]
Well, first you need to measure schema load overhead by calling `GetSchema` API. I have doubts it will take a lot of time for SQL Server if you don't load stored procedures.

I would suspect that performance issue lies in T4 pipeline as Visual Studio does a lot of stuff here and I too noticed that sometimes T4 generation hangs.

Anyway:
1. feature like that is not on our radar and we are full with other tasks, but we can accept PR
2. with release of [cli-based](https://www.nuget.org/packages/linq2db.cli) scaffold tool we've stopped active development for T4 templates. I would recommend you to try it

**Reply by dipique** [reply]
You were right about it being almost entirely (> 99%) due to stored procedures. I've managed to pull apart the SP's into a separate model which has removed most of the pain.

I look forward to using the scaffolding tool but right now introducing .net core code into our codebase is not an option. Soon, hopefully :) Thanks for your help.

**Reply by MaceWindu** [reply]
You don't need to migrate your project to .net core to use dotnet tool - it's standalone application, which doesn't depend on user's project
